File: stool/engines/generic.py
"""通用兜底插件：未知引擎时给出排查建议与基础文件浏览。"""
from __future__ import annotations

import os
import subprocess
from pathlib import Path

from ..core.plugin import Detection, OpResult
from ..core.util import ProgressFn
from .base import EngineBase

# 常见封包扩展名 → 提示
KNOWN_EXT = {
    ".xp3": "KiriKiri", ".rpa": "Ren'Py", ".rgssad": "RPG Maker XP",
    ".rgss2a": "RPG Maker VX", ".rgss3a": "RPG Maker VX Ace",
    ".wolf": "Wolf RPG", ".nsa": "NScripter", ".pck": "Godot",
    ".asar": "Electron", ".aqp": "AdvSys/ARCG", ".pac": "通用封包",
    ".arc": "通用封包", ".dat": "通用数据", ".pak": "通用封包",
    ".bin": "通用二进制", ".pfs": "Ethornell/BGI", ".scx": "ScenePlayer",
    ".mjo": "Mink", ".int": "Interlude", ".dpm": "Donut", ".noa": "Tactics",
}


class GenericPlugin(EngineBase):
    id = "generic"
    name = "未知引擎（兜底）"
    priority = -100

    def detect(self, root: Path) -> Detection:
        return Detection(score=10, evidence=["未匹配到已知引擎特征"])

    def describe(self, root: Path) -> str:
        hits = {}
        for p in root.rglob("*"):
            if p.is_file() and p.suffix.lower() in KNOWN_EXT:
                hits.setdefault(KNOWN_EXT[p.suffix.lower()], []).append(p.name)
        if hits:
            return "疑似: " + "; ".join(f"{k}({v[0]})" for k, v in list(hits.items())[:5])
        return "未发现特征封包，建议用 GARbro 打包器逐个尝试"

    def do_extract(self, root: Path, out_dir: Path, options: dict, progress: ProgressFn) -> OpResult:
        cfg_root = root
        garbro = load_garbro()
        if garbro and Path(garbro).exists():
            r = subprocess.run([garbro, str(cfg_root), "-o", str(out_dir)],
                               capture_output=True, text=True)
            return OpResult(True, f"GARbro 批量提取完成 → {out_dir}" if r.returncode == 0
                            else f"GARbro 失败: {r.stderr[-300:]}")
        return OpResult(False, "未知格式：请在设置页配置 GARbro CLI 路径，"
                               "或参照报告中的“未知格式排查顺序”人工处理")


def load_garbro():
    from ..core.paths import load_config
    return load_config().get("garbro")

File: stool-rs/src/engines/generic.rs
//! 未知引擎兜底插件：特征提示 + GARbro 委派。

use std::path::Path;
use std::process::Command;

use super::scan::ScanCtx;
use super::{Ctx, Detection, Engine, Op, OpOutcome};
use crate::settings;

/// 扩展名 → 疑似引擎。用于"未识别"时给出人话提示（不再只是"未知"）。
const KNOWN_EXT: &[(&str, &str)] = &[
    ("xp3", "KiriKiri"), ("rpa", "Ren'Py"), ("rgssad", "RPG Maker XP"),
    ("rgss2a", "RPG Maker VX"), ("rgss3a", "RPG Maker VX Ace"), ("wolf", "Wolf RPG"),
    ("nsa", "NScripter"), ("pck", "Godot"), ("asar", "Electron"), ("aqp", "AdvSys/ARCG"),
    ("arc", "Ethornell/Majiro（同名 .arc，需看 exe 判断）"),
    ("pfs", "Ethornell/BGI"), ("pfs2", "Artemis"),
    ("mjo", "Majiro"), ("dpm", "Donut"),
    ("ypf", "YU-RIS"), ("int", "CatSystem2"), ("ldb", "RPG Maker 2000/2003"),
    ("pak", "Siglus/Unreal/通用"), ("utoc", "Unreal Engine 5"), ("ucas", "Unreal Engine 5"),
    ("ald", "AliceSoft"), ("npk", "Nitroplus"), ("npa", "Nitroplus"),
    ("prt", "LiveMaker"), ("grp", "LiveMaker"), ("love", "LÖVE/Love2D"),
    ("lpk", "Purple"), ("ypf2", "YU-RIS"),
];

pub struct GenericPlugin;

impl Engine for GenericPlugin {
    fn id(&self) -> &'static str {
        "generic"
    }
    fn name(&self) -> &'static str {
        "未知引擎（兜底）"
    }
    fn priority(&self) -> i32 {
        -100
    }
    fn detect_scan(&self, _scan: &ScanCtx) -> Detection {
        let mut d = Detection::new(self.id(), self.name());
        d.hit(0, "未匹配到任何已收录的引擎特征");
        d.note("请在设置页配置 GARbro 后点“解包”，或参照“未知格式排查顺序”人工处理");
        d
    }
    fn capabilities(&self) -> Vec<Op> {
        vec![Op::Extract]
    }
    fn describe(&self, root: &Path) -> String {
        let scan = ScanCtx::build(root);
        let mut hits: Vec<String> = Vec::new();
        for (ext, count) in scan.top_exts(40) {
            if let Some((_, eng)) = KNOWN_EXT.iter().find(|(k, _)| *k == ext) {
                if !hits.iter().any(|h| h.starts_with(eng.split('（').next().unwrap_or(eng))) {
                    hits.push(format!("{eng}(.{ext} × {count})"));
                }
            }
            if hits.len() >= 6 {
                break;
            }
        }
        let head = scan.summary();
        if hits.is_empty() {
            format!("{head}；未发现特征封包，建议用 GARbro 逐个尝试")
        } else {
            format!("{head}；疑似: {}", hits.join("; "))
        }
    }
    fn extract(&self, ctx: &Ctx) -> OpOutcome {
        garbro_extract(ctx)
    }
}

/// GARbro 委派解包（兜底插件与盲区识别器共用）。
pub fn garbro_extract(ctx: &Ctx) -> OpOutcome {
    let cfg = settings::load();
    match settings::external_tool(&cfg, "garbro") {
        Some(garbro) => match Command::new(garbro)
            .args([ctx.root.to_string_lossy().as_ref(), "-o", ctx.out_dir.to_string_lossy().as_ref()])
            .output()
        {
            Ok(o) if o.status.success() => OpOutcome::ok(format!("GARbro 批量提取完成 → {}", ctx.out_dir.display())),
            Ok(o) => {
                let msg = if o.stderr.is_empty() { o.stdout } else { o.stderr };
                OpOutcome::fail(format!("GARbro 失败: {}", String::from_utf8_lossy(&msg)))
            }
            Err(e) => OpOutcome::fail(format!("GARbro 调用失败: {e}")),
        },
        None => OpOutcome::fail(
            "该格式没有内置解析器：请在设置页配置 GARbro CLI 路径后重试，或参照“未知格式排查顺序”人工处理",
        ),
    }
}
